addons/account/models/account_move.py

```diff
@@ -503,12 +503,40 @@ def _onchange_partner_id(self):
 
         # Find the new fiscal position.
         delivery_partner_id = self._get_invoice_delivery_partner_id()
+        # Keep track of whether the fiscal position has changed
+        previous_fiscal_position_id = self.fiscal_position_id
         self.fiscal_position_id = self.env['account.fiscal.position'].get_fiscal_position(
             self.partner_id.id, delivery_id=delivery_partner_id)
-        self._recompute_dynamic_lines()
+
+        # Update the taxes if the fiscal position has changed
+        if previous_fiscal_position_id != self.fiscal_position_id:
+            self._onchange_fiscal_position_id()
+        else:
+            self._recompute_dynamic_lines()
+
         if warning:
             return {'warning': warning}
 
+    @api.onchange('fiscal_position_id')
+    def _onchange_fiscal_position_id(self):
+        for line in self.line_ids:
+            # Only update lines with products that have defined taxes
+            if (
+                not line.product_id.taxes_id
+                and self.move_type in ('out_invoice', 'out_refund')
+                or (
+                    not line.product_id.supplier_taxes_id
+                    and self.move_type in ('in_invoice', 'in_refund')
+                )
+                or line.display_type in ('line_section', 'line_note')
+            ):
+                continue
+            line.tax_ids = line._get_computed_taxes()
+            line.account_id = line._get_computed_account()
+            line.recompute_tax_line = True
```

Comment:
Maybe I miss something but I'm pretty sure this loop should be something like that:
```
for line in self.line_ids:
    if not line.move_id.is_invoice(include_receipts=True) or line.display_type or line.exclude_from_invoice_tab:
        continue

    new_taxes = line._get_computed_taxes()
    if set(line.tax_ids) != set(new_taxes):
        line.tax_ids = new_taxes
        line.recompute_tax_line = True
    line.account_id = line._get_computed_account()
```

Reply:
@smetl don't forget to add the language in your code snippets ;)

Reply:
ah mostly, the set comparison is a good idea, but the conditional is necessarily complex, we need to check the product 

has sales taxes if its an out invoice _or_
has supplier taxes if its an in invoice

to ensure that the original tax can be returned to. There's a scenario where a fiscal position maps tax A -> tax B, but there's no fiscal position mapping tax B -> tax A, so when a user changes customer, the taxes get changed, but when the switch back to the original partner, the taxes do not return to the original taxes.